### PR TITLE
Retry models that may beat the top model based on the current epsilon

### PR DESCRIPTION
**Pull request by RusticLuftig**

### CODE REVIEW

File: neurons/validator.py [hunk 1]
```diff
@@ -519,6 +440,132 @@ def update_models(self):
...
+        # Find the most recent successful eval.
+        last_successful_eval = None
+        for eval_result in reversed(eval_history):
+            if eval_result.avg_loss != math.inf:
+                last_successful_eval = eval_result
+                break
+
+        if last_successful_eval:
+            # If this model had worse loss than the top model during the last eval, no need to retry.
+            if last_successful_eval.score > last_successful_eval.winning_model_score:
```
Comment (Sid-Data-Universe) [inline neurons/validator.py:457]: This rename is a little awkward since we are saying if the 'score' is higher to not retry.

**Review by Sid-Data-Universe — DISMISSED** [review]

**Comment by Sid-Data-Universe** [comment]
(Also LGTM for this overall part of the code. Assuming the code to write to the new data structure in the eval loop is also included in this or another PR)

File: neurons/validator.py (hunk 1 above)
Comment (RusticLuftig) [reply neurons/validator.py]: Ack. Added a comment to help make this clearer

**Comment by RusticLuftig** [comment]
> (Also LGTM for this overall part of the code. Assuming the code to write to the new data structure in the eval loop is also included in this or another PR)

That was a little more work than expected due to the clean-up in log_step. That method is such a mess that we should really clean up at some point...

**Pull request merged**